authui/src/loading.ts

```diff
@@ -21,6 +21,18 @@ export function disableAllButtons(): () => void {
   };
 }
 
+// displayButtonLoadingState remembers the content of button.
+// It show loading state on button and returns a function to revert to original content.
+export function displayButtonLoadingState(el: HTMLButtonElement): () => void {
+  const original: string = el.innerHTML;
+  el.innerHTML =
+    '<span class="primary-btn__loading-icon material-icons animate-spin">progress_activity</span>';
```

Comment:
1. I think we should decouple this with XHRSubmitFormController.
2. Replacing the original content with the loading spinner could cause layout bouncing because their content size is different.

My proposed solution:

- Define a new controller named `loadable-button`.
  - Usage: `data-controller="loadable-button"`
  - On connect(): Inject the `<span class="loading-spinner material-icons">` element immediately.
  - `loading-spinner` should be a class which by default hide the element if the parent is not loading.
  ```css
  :not(.loading) > .loading-spinner {
    display:  none;
  }
  ```
- In `displayButtonLoadingState`, you should only add a class `loading` to the button element. Do not mutate the children. The returned function (dismissButtonLoadingState) should only remove the class from classList of the same element.
- Overlay the spinner on top of the content, instead of replacement.
  ```css
  .loading {
    display: relative;
  }
  .loading :not(.loading-spinner) {
    visibility: hidden; /* Hide all things except the spinner */
  }
  .loading-spinner {
    @apply absolute inset-0; /* Covers the whole button and center align */
  }
  ```

Reply:
Create a controller solely for injecting the loading spinner but not responsible for controlling the loading state display looks quite odd to me. 

Seems like we don't actually need a connector. Can just do something like the follow? Which is similar to what I do at the very first place? https://github.com/authgear/authgear-server/pull/3654#discussion_r1461778200

Just we delegate the loading state visibility control to XHRSubmitFormController
```html
<button class="button">
  this is a button
  <span class="loading-spinner"></span>
</button>
```

Reply:
My thought is we don't want the user of the primary button to declare the loading state ui each time they use the button. i.e. `<span class="loading-spinner material-icons">` should be reused in all reloadable button, because we expect the loading ui is same.

Reply:
Then I think define a template would be a more proper solution?

```html
{{ define "loading_spinner" }}
 <span class="loading-spinner"></span>
{{ end }}
```

Reply:
I feel like it is more complicated.

In my understanding, `loadable-button` is one component, and the css `loadable-button.css` handles the display and hide logic of the spinner.

And using controller has an advantage that you can query the element by selector `data-controller~="loadable-button"`. So you don't need the additional attribute `data-button-loading-state`.

Reply:
Should we just move primary button be moved to html template to explicitly declare the loading spinner. `loadable` controller will solely control the class list 🤔

Reply:
I think we can do it with css pseudo element.

1. Make the class `.material-icons` a utiliy.
  ```css
  @layer utilities {
    .material-icons { ... }
  }
  ```
2. Define a class with the pseudo element using the above utility.
  ```css
  .loadable {
    position: relative;

    &::after {
      @apply material-icons animate-spin;
      content: "progress_activity";
    }

    &:not(.loading)::after {
      @apply hidden;
    }
  }
  ```

In this way we don't need template nor controller, what do you think?

Reply:
In portal, we decouple the state of loading and the presentation of loading. We have src/hook/loading to allow components at any levels to declare there is something loading. We also allow components at any levels to read the current loading state. Would it be too overkill if we introduce a similar abstraction here? 

In my mind, We can have a LoadingController and XHRSubmitFormController mounted at very top level in the HTML document. XHRSubmitFormController can ask LoadingController to become loading, either via [event](https://stimulus.hotwired.dev/reference/controllers#cross-controller-coordination-with-events) or [direct invocation](https://stimulus.hotwired.dev/reference/controllers#directly-invoking-other-controllers). The missing piece is how do we assign an element responsible for the presentation of loading. Personally, I think the API is the most intuitive if I can simply declare which element is responsible for the presentation of loading in our templates. For example

```
<button class="primary-btn" data-loading-presentation-target>Continue</button>
```